ledger/src/blockstore.rs

```diff
@@ -1532,22 +1564,86 @@ impl Blockstore {
         shred_index < slot_meta.consumed || data_index.contains(shred_index)
     }
 
-    fn get_data_shred_from_just_inserted_or_db<'a>(
+    /// Finds the corresponding shred at `shred_id` in the just inserted
+    /// shreds or the backing store. Panics if there is no shred.
+    fn get_shred_from_just_inserted_or_db<'a>(
         &'a self,
         just_inserted_shreds: &'a HashMap<ShredId, Shred>,
-        slot: Slot,
-        index: u64,
+        shred_id: ShredId,
     ) -> Cow<'a, Vec<u8>> {
-        let key = ShredId::new(slot, u32::try_from(index).unwrap(), ShredType::Data);
-        if let Some(shred) = just_inserted_shreds.get(&key) {
-            Cow::Borrowed(shred.payload())
-        } else {
+        let (slot, index, shred_type) = shred_id.unpack();
+        match (just_inserted_shreds.get(&shred_id), shred_type) {
+            (Some(shred), _) => Cow::Borrowed(shred.payload()),
             // If it doesn't exist in the just inserted set, it must exist in
             // the backing store
-            Cow::Owned(self.get_data_shred(slot, index).unwrap().unwrap())
+            (_, ShredType::Data) => Cow::Owned(
+                self.get_data_shred(slot, u64::from(index))
+                    .unwrap()
+                    .unwrap_or_else(|| {
+                        panic!("{} {} {:?} must be present!", slot, index, shred_type)
+                    }),
+            ),
+            (_, ShredType::Code) => Cow::Owned(
+                self.get_coding_shred(slot, u64::from(index))
+                    .unwrap()
+                    .unwrap_or_else(|| {
+                        panic!("{} {} {:?} must be present!", slot, index, shred_type)
+                    }),
+            ),
         }
     }
 
+    /// Returns true if there is no merkle root conflict between
+    /// the existing `merkle_root_meta` and `shred`
+    ///
+    /// Otherwise return false and if not already present, add duplicate proof to
+    /// `duplicate_shreds`.
+    fn perform_merkle_check(
+        &self,
+        just_inserted_shreds: &HashMap<ShredId, Shred>,
+        slot: Slot,
+        merkle_root_meta: &MerkleRootMeta,
+        shred: &Shred,
+        duplicate_shreds: &mut Vec<PossibleDuplicateShred>,
+    ) -> bool {
+        let new_merkle_root = shred.merkle_root().ok();
```

Comment:
Why don't we check for the `feature_set::merkle_conflict_duplicate_proofs` feature activation here in this function?

Reply:
I also considered this as an option - since this is code is temporary until the feature flag is turned on I went with this approach to keep the PR small.

There are 183 calls (mostly all testing code) for `blockstore.insert_shreds` that we would need to be altered and then removed on feature flag cleanup.

 I figured it would be easiest to check the feature flag in `window_service` and once the feature flag has been activated move the `blockstore.store_duplicate_slot` from `window_service` to `blockstore.insert_shreds`

Reply:
fine, if there is no correctness concern.
e.g. when blockstore returns a duplicate proof it will stop processing there and ignores other variants of duplicate proofs.
But if window-service ignores that because the feature is not activated, then it is possible that the other variants of duplicates proofs were also present but not processed.

Reply:
since we don't insert into blockstore until window service, other variants of duplicate proofs will not be precluded from being processed if the feature flag is not present. 

additionally because they are separate enum variants, multiple signals can be sent to window service if a block violates multiple duplicate conditions.